File: src/main/java/space/ryzhenkov/fabric2discord/mixins/ServerPlayerEntityMixin.java
package space.ryzhenkov.fabric2discord.mixins;

import com.mojang.authlib.GameProfile;
import net.minecraft.entity.damage.DamageSource;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.server.world.ServerWorld;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;

import java.util.Objects;

@Mixin(ServerPlayerEntity.class)
public abstract class ServerPlayerEntityMixin extends PlayerEntity {
    public ServerPlayerEntityMixin(ServerWorld world, GameProfile profile) {
        super(world, world.getSpawnPos(), world.getSpawnAngle(), profile);
    }

    @Inject(method = "onDeath(Lnet/minecraft/entity/damage/DamageSource;)V", at = @At("HEAD"))
    private void onDeath(DamageSource source, CallbackInfo ci) {
        space.ryzhenkov.fabric2discord.ktmixins.ServerPlayerEntityMixin.INSTANCE.onDeath(Objects.requireNonNull(this.getCommandSource().getPlayer()), source, getDamageTracker());
    }

    @Inject(method = "worldChanged", at = @At("TAIL"))
    private void worldChanged(ServerWorld origin, CallbackInfo ci) {
        space.ryzhenkov.fabric2discord.ktmixins.ServerPlayerEntityMixin.INSTANCE.worldChanged(Objects.requireNonNull(this.getCommandSource().getPlayer()), origin);
    }
}

File: src/main/kotlin/space/ryzhenkov/fabric2discord/ktmixins/ServerPlayerEntityMixin.kt
package space.ryzhenkov.fabric2discord.ktmixins

import net.minecraft.entity.damage.DamageSource
import net.minecraft.entity.damage.DamageTracker
import net.minecraft.server.network.ServerPlayerEntity
import net.minecraft.server.world.ServerWorld
import space.ryzhenkov.fabric2discord.config.ConfigAPI
import space.ryzhenkov.fabric2discord.utils.MessageUtils

object ServerPlayerEntityMixin {
    fun onDeath(player: ServerPlayerEntity, source: DamageSource, tracker: DamageTracker) {
        if (!ConfigAPI.messages.playerDeath.enabled) return
        if (ConfigAPI.general.ids.getLogChannelOrNull() == null) return

        val replacements = hashMapOf<String, String>()
        replacements["death_message"] = tracker.deathMessage.string
        replacements["death_by"] = source.name

        MessageUtils.sendEmbedMessage(ConfigAPI.general.ids.getLogChannelOrNull()) {
            MessageUtils.getEmbedMessage(ConfigAPI.messages.playerDeath, player, replacements)
        }
    }

    fun worldChanged(player: ServerPlayerEntity, origin: ServerWorld) {
        if (!ConfigAPI.messages.playerDimension.enabled) return
        if (ConfigAPI.general.ids.getLogChannelOrNull() == null) return

        val replacements = hashMapOf<String, String>()
        replacements["world_origin"] = origin.registryKey.value.toString()
        replacements["world_origin_id"] = origin.registryKey.value.path

        MessageUtils.sendEmbedMessage(ConfigAPI.general.ids.getLogChannelOrNull()) {
            MessageUtils.getEmbedMessage(ConfigAPI.messages.playerDimension, player, replacements)
        }
    }
}